package/tvheadend-app-osmc/files/DEBIAN/postinst

```diff
@@ -1,33 +1,182 @@
-#!/bin/bash
+#! /bin/bash
 
 if [ "$1" = "configure" ]; then
 
 	systemctl daemon-reload
+#set -x
+	# going to read existing config files even if installing
+	# unless version is 4.2 when we can trust tvh to do it
+	EXISTINGVER=$(egrep "full_?version" /home/osmc/.hts/tvheadend/config | sed 's/.*: \+\"\(.*\)\".*/\1/' 2> /dev/null)
+	
+	if $(dpkg --compare-versions "$2" lt "4.2.0") || [[ "$EXISTINGVER" < "4.2.0" ]] \
+		|| [ -e "/var/run/${DPKG_MAINTSCRIPT_PACKAGE}_install" ]; then
+#	if ((1)); then
+		# check for existing user called osmc
+		if OSMCUSERFILES=$(ls -t /home/osmc/.hts/tvheadend/accesscontrol/* 2> /dev/null); then
+			OSMCUSERFILES=($(grep -l "\"username\": \"osmc\"" $OSMCUSERFILES  2> /dev/null))
+			#echo ${OSMCUSERFILES[@]}
+			OSMCUSERFILE=${OSMCUSERFILES[0]}
+		fi
+		if OSMCPASSFILES=$(ls -t /home/osmc/.hts/tvheadend/passwd/*  2> /dev/null); then
+			OSMCPASSFILES=($(grep -l "\"username\": \"osmc\"" $OSMCPASSFILES  2> /dev/null))
+			OSMCPASSFILE=${OSMCPASSFILES[0]}
+		fi
 
-	if [ ! -f /home/osmc/.hts/tvheadend/accesscontrol/1 ] && [ -e "/var/run/${DPKG_MAINTSCRIPT_PACKAGE}_install" ]; then
+		# save any passwords
+		if [ -n "$OSMCUSERFILE" ] && [ -z $OSMCPASSFILE ]; then
+			OSMCUSERPASS=$(grep "\"password2\":" "$OSMCUSERFILE")
+		else
+			OSMCUSERPASS=$(grep "\"password2\":" "$OSMCPASSFILE")
+		fi
+		OSMCUSERPASS=${OSMCUSERPASS%,}
+
+		#echo userfile is: $OSMCUSERFILE
+		#echo password file is: $OSMCPASSFILE
+		#echo password: $OSMCUSERPASS
+		#echo current version: $EXISTINGVER
 
-		echo "Creating TVHeadend default access control file"
 		mkdir -p /home/osmc/.hts/tvheadend/accesscontrol
+		mkdir -p /home/osmc/.hts/tvheadend/passwd
 
-		cat <<- 'EOF' > /home/osmc/.hts/tvheadend/accesscontrol/1
-		{
-			"enabled": 1,
-		        "username": "osmc",
-		        "password": "osmc",
-		        "comment": "Default access entry",
-		        "prefix": "0.0.0.0/0,::/0",
-		        "streaming": 1,
-		        "dvr": 1,
-		        "dvrallcfg": 1,
-		        "webui": 1,
-		        "admin": 1,
-		        "id": "1"
+		function addparameter(){
+			if ! grep "$2" "$1" > /dev/null; then
+				LINE=$(sed -n '$=' "$1")
+				#echo $LINE
+				while [[ $(sed -n $LINE's/.*[]\"].*/found/p' "$1") != 'found' ]] && [[ $LINE > 0 ]] ; do
+					LINE=$(($LINE -1))
+				done
+				if [[ $LINE == 0 ]]; then
+					echo Problem with user file $1 - edit or delete it
+					exit 1
+				fi
+				sed -i $LINE'{
+					s/.*[^,]$/&,/
+					a'"$2"'
+				}' "$1"
+			fi
 		}
+
+		if [ -z "$OSMCUSERFILE" ]; then
+			OSMCUSERFILE=/home/osmc/.hts/tvheadend/accesscontrol/ef332059f7bb54d72ce2e028f71e1540
+
+			echo "Creating OSMC default access control file"
+
+			if [[ "$EXISTINGVER" < "4.2.0" ]];then
+				# make a 4.0.9 version user file (as posted by Jaroslav in tvh forum)
+				cat <<- 'EOF' > "$OSMCUSERFILE"
```

Comment:
This caters for the situation where someone has not used the default username in their 4.0.9 install.  So we at least have one user which works in the webui (in particular in the wizard) with a predictable name and password.  I thought people with complicated setups like lots of usernames would be best to sort those out manually - they may well have different permissions, streaming/recording profiles etc which the hacks may affect.

So I don't propose to cycle through all users and attempt to fettle them for migration.  TVH will migrate them but their configs will have to be checked manually in the webui.